File: perf/benchmarks/micro/fibonacci/fibonacci.py
# Fibonacci — recursive function-call micro-benchmark.
#
# Computes fib(N) recursively then prints CHECKSUM and ELAPSED_MS.
# Argument: N (default 32).

import sys
import time


def fib(n: int) -> int:
    if n < 2:
        return n
    return fib(n - 1) + fib(n - 2)


def main() -> None:
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 32
    sys.setrecursionlimit(max(1000, n + 100))
    t0 = time.perf_counter_ns()
    result = fib(n)
    elapsed_ns = time.perf_counter_ns() - t0
    print(f"CHECKSUM:{result}")
    print(f"ELAPSED_MS:{elapsed_ns / 1_000_000:.6f}")


if __name__ == "__main__":
    main()

File: perf/benchmarks/micro/fibonacci/fibonacci.cpp
// Fibonacci — recursive function-call micro-benchmark (C++).
#include <chrono>
#include <cstdint>
#include <cstdio>
#include <cstdlib>

static int64_t fib(int n) {
    if (n < 2) return n;
    return fib(n - 1) + fib(n - 2);
}

int main(int argc, char** argv) {
    int n = (argc > 1) ? std::atoi(argv[1]) : 32;
    auto t0 = std::chrono::steady_clock::now();
    int64_t result = fib(n);
    auto t1 = std::chrono::steady_clock::now();
    double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
    std::printf("CHECKSUM:%lld\n", static_cast<long long>(result));
    std::printf("ELAPSED_MS:%.6f\n", ms);
    return 0;
}
